p2p/peerResend.go

```diff
@@ -0,0 +1,99 @@
+package p2p
+
+import (
+	"crypto/sha1"
+	"sync"
+	"time"
+)
+
+type PeerResend struct {
+	mtx      sync.RWMutex
+	buckets  []*PRBucket
+	pc       int
+	interval time.Duration
+	stopper  chan interface{}
+}
+
+type PRBucket struct {
+	mtx  sync.RWMutex
+	data map[[sha1.Size]byte]bool
+}
+
+func (b *PRBucket) Has(hash [sha1.Size]byte) bool {
+	b.mtx.RLock()
+	defer b.mtx.RUnlock()
+	return b.data[hash]
+}
+
+func (b *PRBucket) Add(hash [sha1.Size]byte) {
+	b.mtx.Lock()
+	defer b.mtx.Unlock()
+	b.data[hash] = true
+}
+
+func NewPeerResend(buckets int, interval time.Duration) *PeerResend {
+	pr := new(PeerResend)
+	pr.interval = interval
+	pr.buckets = make([]*PRBucket, buckets)
+	for i := 0; i < buckets; i++ {
+		pr.buckets[i] = newPRBucket()
+	}
+	pr.stopper = make(chan interface{}, 1)
+	go pr.Cleanup()
+	return pr
+}
+
+func newPRBucket() *PRBucket {
+	b := new(PRBucket)
+	b.data = make(map[[sha1.Size]byte]bool)
+	return b
+}
+
+func (pr *PeerResend) Has(hash [sha1.Size]byte) bool {
+	pr.mtx.RLock()
+	defer pr.mtx.RUnlock()
+
+	for i := 0; i < len(pr.buckets); i++ {
+		if pr.buckets[pr.ptr(i)].Has(hash) {
+			return true
+		}
+	}
+	return false
+}
+
+func (pr *PeerResend) Add(hash [sha1.Size]byte) {
+	pr.mtx.RLock()
+	defer pr.mtx.RUnlock()
```

Comment:
same here, I am not convinced using defer when you don't really need it. My understanding is that it has a minimal overhead, and also slightly decrease readability by having a non sequential order of operations written